dd-java-agent/instrumentation/grpc-1.5/src/main/java/datadog/trace/instrumentation/grpc/server/MethodHandlersInstrumentation.java

```
@@ -0,0 +1,60 @@
+package datadog.trace.instrumentation.grpc.server;
+
+import static datadog.trace.agent.tooling.bytebuddy.matcher.NameMatchers.nameEndsWith;
+import static datadog.trace.bootstrap.debugger.spanorigin.CodeOriginInfo.entry;
+import static net.bytebuddy.matcher.ElementMatchers.isConstructor;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter.ForTypeHierarchy;
+import datadog.trace.agent.tooling.InstrumenterModule;
+import java.lang.reflect.Method;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(InstrumenterModule.class)
+public class MethodHandlersInstrumentation extends InstrumenterModule.Tracing
+    implements ForTypeHierarchy {
+  private static final ElementMatcher<TypeDescription> METHOD_HANDLERS =
+      nameEndsWith("$MethodHandlers");
+
+  public MethodHandlersInstrumentation() {
+    super("grpc-server-code-origin");
+  }
+
+  @Override
+  public String hierarchyMarkerType() {
+    return "io.grpc.MethodDescriptor";
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return METHOD_HANDLERS;
+  }
+
+  @Override
+  public void methodAdvice(MethodTransformer transformer) {
+    transformer.applyAdvice(
+        isConstructor().and(takesArguments(2)),
+        MethodHandlersInstrumentation.class.getName() + "$BuildAdvice");
+  }
+
+  public static class BuildAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnter(@Advice.Argument(0) Object serviceImpl) {
+      Class<?> serviceClass = serviceImpl.getClass();
```

I'll also have wrapped this under a big try catch throwable. It's true that exceptions are suppressed by bytebuddy but it can pollute then our telemetry

---

Would you just swallow the exception?  I had that wrapped at one point but I ended up removing it because while testing it was getting exceptions thrown and silently caught but all I saw was that my instrumentation "wasn't working." If the exception gets thrown, at least the surrounding infrastructure would log that failure in ways I can't from this method.

---

you can log it on debug logs so if a customer has an issue and provide those logs we can know. However, letting the error be catched by the default handler will result this error reporter as a telemetry and we monitor it as an unexpected error

---

added the try/catch and a note to $futureDev